segmentation.py

```diff
@@ -0,0 +1,575 @@
+                        consider. After each update, pruning will take place to
+                        limit the number of hypotheses. If set to ``None``, no
+                        pruning will NOT take place after updates, however,
+                        pruning can be initiated manually by calling
+                        :py:meth:`.trim`.
+
+    Raises:
+        Exception: If the any of the inputs are an incorrect type.
+
+    """
+
+    def __init__(self, mu=None, omega=None, sigma=None, eta=None,
+                 alg='sumprod', ratefun=0.1, featfun=None, minprob=1.0e-6,
+                 maxhypot=20):
+
+        # The inference algorithm must be either sum-product or sum-product.
+        if alg.lower() not in ['sumprod', 'maxprod']:
+            msg = "The input 'alg' must be either 'sumprod' or 'maxprod'."
+            raise Exception(msg)
+        else:
+            self.__alg__ = alg.lower()
+
+        # Store number of dimensions in the predictor (independent/input
+        # variable) and response (dependent/output variable) variables.
+        self.__m = None
+        self.__n = None
+
+        # Allocate variables for matrix variate, normal inverse gamma
+        # distributions.
+        self.__mu = None
+        self.__omega = None
+        self.__sigma = None
+        self.__eta = None
+
+        # Set prior for the location parameter.
+        if mu is not None:
+            self.__mu = mu
+
+        # Set prior for the scale parameter.
+        if omega is not None:
+            self.__omega = omega
+
+        # Set prior for the dispersion/noise parameter.
+        if sigma is not None:
+            self.__sigma = sigma
+
+        # Set prior for the shape parameter.
+        if eta is not None:
+            self.__eta = eta
+
+        # Ensure algorithm initialises on first call to update.
+        self.__initialised = False
+
+        # If 'maxhypot' is set to none, no hypotheses will be trimmed.
+        if maxhypot > 0 or not None:
+            self.__maximum_hypotheses = maxhypot
+        else:
+            msg = "The input 'maxhypot' must be an integer greater than zero."
+            raise Exception(msg)
+
+        if minprob > 0:
+            self.__minimum_probability = minprob
+        else:
+            msg = "The input 'minprob' must be a float greater than zero."
+            raise Exception(msg)
+
+        # Allocate variables for tracking segments.
+        self.__hypotheses = list()
+        self.__index = list()
+        self.__probabilities = list()
+
+        self.__featfun = featfun if callable(featfun) else lambda x: x
+        self.__ratefun = ratefun if callable(ratefun) else lambda x: ratefun
+
+    def __initialise_algorithm(self, m, n):
+        """Initialise the Bcdm algorithm."""
+
+        # Ensure input dimensions are consistent.
+        if self.__m is None:
+            self.__m = m
+        elif self.__m != m:
+            msg = 'Expected %i dimensions in the predictor variable.' % m
```

Comment:
Probably! Yes. Sorry. Again, I think I did this early on before I had a view to contribute. I simply liked the older notation better since it can be more compact. But 'format' is more flexible.